### PR TITLE
feat: use sarama lib for internal message publishing

### PR DESCRIPTION
**Pull request by turip**
## Overview

This patch uses the sarama lib for internal message publishing. This patch also adds support for exposing the sarama metrics using the metrics bridge.

# Events

Sarama (any most probably other projects) rely on the [http://github.com/rcrowley/go-metrics](go-metrics) package, numerous connectors exist
for the package, however there seem to be no opentelemetry one.

Given the package only supports periodic scraping, it's better to wrap the metric types of go-metrics so that we can send the raw events to OpenTelemetry. The existing event interface of go-metrics is quite limited compared to OpenTelemetry esp considering the usage in Sarama lib, so right now:

- Context is context.Background() for OpenTelemetry calls
- Errors for metric registration is only logged

Example metrics:

```
# TYPE sarama_publisher_response_size histogram
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="0"} 0
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="5"} 0
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="10"} 0
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="25"} 0
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="50"} 0
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="75"} 2
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="100"} 4
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="250"} 4
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="500"} 6
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="750"} 6
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="1000"} 6
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="2500"} 6
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="5000"} 6
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="7500"} 6
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="10000"} 6
sarama_publisher_response_size_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="+Inf"} 6
sarama_publisher_response_size_sum{otel_scope_name="openmeter.io/backend",otel_scope_version=""} 880
sarama_publisher_response_size_count{otel_scope_name="openmeter.io/backend",otel_scope_version=""} 6
# HELP sarama_publisher_response_size_for_broker_1 
# TYPE sarama_publisher_response_size_for_broker_1 histogram
sarama_publisher_response_size_for_broker_1_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="0"} 0
sarama_publisher_response_size_for_broker_1_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="5"} 0
sarama_publisher_response_size_for_broker_1_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="10"} 0
sarama_publisher_response_size_for_broker_1_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="25"} 0
sarama_publisher_response_size_for_broker_1_bucket{otel_scope_name="openmeter.io/backend",otel_scope_version="",le="50"} 0
```

### CODE REVIEW

File: internal/watermill/driver/kafka/metrics/adapter.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
+package metrics
```
Comment (turip) [inline internal/watermill/driver/kafka/metrics/adapter.go:1]: @sagikazarmark this is not pretty, but I couldn't find an opensource bridge/adapter supporting this. We can consider putting this into a separate opensource package with an extended review, WDYT?

File: internal/watermill/driver/kafka/publisher.go [hunk 2]
```diff
@@ -1,63 +1,108 @@
...
+		return errors.New("metric meter is required")
+	}
 	return nil
 }
 
-func (p *Publisher) Close() error {
-	p.producer.Close()
-	return nil
-}
+func NewPublisherFromOMConfig(in PublisherOptions) (*kafka.Publisher, error) {
```
Comment (GAlexIHU) [inline internal/watermill/driver/kafka/publisher.go:50]: nit: `NewPublisherFromOMConfig` => `NewPublisher`

**Review by GAlexIHU — APPROVED** [review]
👍

File: openmeter/watermill/driver/kafka/driver.go [hunk 3]
```diff
@@ -13,11 +13,11 @@ const (
 )
 
 type (
-	Publisher = watermillkafka.Publisher
+	PublisherOptions = watermillkafka.PublisherOptions
 )
 
-func NewPublisher(producer *kafka.Producer) *Publisher {
-	return watermillkafka.NewPublisher(producer)
+func NewPublisherFromOMConfig(in PublisherOptions) (*kafka.Publisher, error) {
```
Comment (sagikazarmark) [inline openmeter/watermill/driver/kafka/driver.go:19]: OMConfig here too

File: internal/watermill/driver/kafka/publisher.go (hunk 2 above)
Comment (turip) [reply internal/watermill/driver/kafka/publisher.go]: Sure, renamed!

File: internal/watermill/driver/kafka/metrics/adapter.go (hunk 1 above)
Comment (turip) [reply internal/watermill/driver/kafka/metrics/adapter.go:1]: Decision we will publish this as a separate package (I will spend some time creating the repo this week or the week after)

**Pull request merged**